presto-main/src/test/java/com/facebook/presto/tracing/testing/TestTestingTracer.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.tracing.testing;
+
+import org.testng.annotations.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+import static com.facebook.airlift.concurrent.Threads.daemonThreadsNamed;
+import static java.util.concurrent.Executors.newFixedThreadPool;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.expectThrows;
+
+public class TestTestingTracer
+{
+    private final TestingTracerProvider tracerProvider;
+    private final ExecutorService executor;
+    private final Random random = new Random();
+
+    public TestTestingTracer()
+    {
+        this.tracerProvider = new TestingTracerProvider();
+        this.executor = newFixedThreadPool(16, daemonThreadsNamed("presto-testing-tracer-threadpool"));
+    }
+
+    @Test
+    public void testAddPoint()
+    {
+        TestingTracer tracer = (TestingTracer) tracerProvider.getNewTracer();
+        List<CompletableFuture> futures = new ArrayList<>();
+        for (int i = 0; i < 10; i++) {
+            CompletableFuture future = new CompletableFuture();
+            executor.submit(() -> {
+                for (int j = 0; j < 20; j++) {
+                    tracer.addPoint("test-point");
+                    try {
+                        Thread.sleep(random.nextInt(5));
```

Comment:
hmmm.... not a big fan of using `sleep`; but given it's millis, it should fine be. QQ: what if random.nextInt(5) gives you 0....?

Reply:
0 is okay here.